Fix Lots of Warnings

wesbland opened the pull request:
## Pull Request Description

This is a grab bag of warning fixes for things not currently covered by the warning checker job. Most of this is related to unused labels that show up when you compile with error checking off (i.e., all of the `MPIR_ERR_CHECK` and `MPIR_Assert` macros disappear), but there are a few others.

Most of these can be reproduced by using `--enable-strict` and `--disable-error-checking`. A few others come from using ICC/ICX instead of GCC.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

wesbland commented:
Ok. I"m done messing with this. It should be ready for review now @travisk1154.

hzhou reviewed (commented):
All the changes look okay to me. But I have to admit I didn't spend much time at all looking into them. I think you can go ahead group the commits by type of changes and type of files, into maybe under 10 or so commits. Meanwhile, I'll ask Ken to also look at them.

wesbland commented:
I squashed a few of these together and grouped the rest together by location in code. Only three are longer than a few lines:

* 4e7f9271c75d - mpi/coll: Suppress unused variable warning
* aeae7eb93e12 - mpl: Cleanup return labels
* fb0fc4628faf - mpl: Fix function definition

wesbland commented:
test:mpich/ch4/most
test:mpich/ch3/most

wesbland commented:
There's a bunch of configs that I used to generate these, but the main thing I did was `--enable-strict` and `--disable-error-checking`. The covers pretty much all of the "unused something" warnings. Some of them came from using ICC/ICX instead of GCC.

At some point, I did give up on getting the labels right because it just gets too complex. That's when I noticed that at some point in the past, we at least said we were going to turn off `-Wunused-labels` in the strict build.

hzhou commented:
>  That's when I noticed that at some point in the past, we at least said we were going to turn off `-Wunused-labels` in the strict build.

We did. But `icc` doesn't have that option. And pgi compilers and suncc also had trouble suppressing it. So at one point I decided to just fix the labels. I wasn't thinking about the `--disable-error-checking`. I guess that makes the issue unfixable, huh?

wesbland commented:
> We did. But `icc` doesn't have that option. And pgi compilers and suncc also had trouble suppressing it. So at one point I decided to just fix the labels. I wasn't thinking about the `--disable-error-checking`. I guess that makes the issue unfixable, huh?

It sounds like that should be resolved by having different sets of strict flags for different compilers rather than reverting to the least common denominator. I have access to ICC (obviously), but not PGI or Sun CC. Is that something you can help with?

wesbland commented:
I potentially have a fix for the different compilers for strict builds coming. Give me a few minutes to test it out.

wesbland commented:
@hzhou Please take a look at the last patch.

hzhou commented on confdb/aclocal_cc.m4:
```diff
@@ -403,6 +403,56 @@ else
+            pac_cv_cc_vendor=pgi
+        elif grep 'Sun Workshop' conftest.txt >/dev/null 2>&1 ; then
+            pac_cv_cc_vendor=sun
+	elif grep 'Sun C' conftest.txt >/dev/null 2>&1 ; then
```
What happened to the indentation? Does your editor fail you? 😄

wesbland replied:
Copy/pasting from the Fortran version with a mix of tabs and spaces failed me...

hzhou commented on confdb/aclocal_cc.m4:
```diff
@@ -403,6 +403,56 @@ else
+    elif test "$CC" = "xlcc" ; then
+        pac_cv_cc_vendor=ibm
+    fi
+fi
```
Missing `clang`?

wesbland replied:
I don't think that's really a problem. It'll stay as "unknown", which means it won't use the GCC-isms. I'll add something anyway...

hzhou commented on confdb/aclocal_cc.m4:
```diff
@@ -508,6 +558,16 @@ if test "$enable_strict_done" != "yes" ; then
+        *)
+            ;;
+    esac
+
```
Yeah, I think this solution is more straightforward. 👍

wesbland commented:
I've addressed the points above.

raffenet commented on src/mpi/coll/algorithms/recexchalgo/recexchalgo.c:
```diff
@@ -272,7 +272,8 @@ int MPII_Recexchalgo_reverse_digits_step2(int rank, int comm_size, int k)
     int pofk = 1, log_pofk = 0;
     int *digit, *digit_reverse;
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno ATTRIBUTE((unused));
+    mpi_errno = MPI_SUCCESS;
```
nitpick: none of the other instances in this commit split the assignment onto a second line

wesbland replied:
Fixed. I learned that I could do that somewhere along the way and didn't backport my simplification.

raffenet commented on src/mpi/coll/transports/gentran/tsp_gentran.c:
```diff
@@ -504,7 +504,7 @@ int MPIR_TSP_sched_sink(MPIR_TSP_sched_t s, int *vtx_id)
 {
     MPII_Genutil_sched_t *sched = s;
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno ATTRIBUTE((unused)) = MPI_SUCCESS;
```
looks like this was leaked from another commit

wesbland replied:
Yep. I've moved it back.

raffenet commented on src/mpid/ch4/src/ch4r_init.c:
```diff
@@ -17,11 +17,8 @@ int MPIDIG_init_comm(MPIR_Comm * comm)
-
-    MPIR_Assert(subcomm_type <= 3);
-    MPIR_Assert(is_localcomm <= 1);
+    MPIR_Assert(MPIR_CONTEXT_READ_FIELD(SUBCOMM, comm->recvcontext_id) <= 3);
+    MPIR_Assert(MPIR_CONTEXT_READ_FIELD(IS_LOCALCOMM, comm->recvcontext_id) <= 1);
```
I think these asserts can just be removed. There used to be a context_id->comm lookup in this places that was removed.

wesbland replied:
Gone.

raffenet commented on src/mpid/ch4/src/ch4r_init.c:
```diff
@@ -31,16 +28,14 @@ int MPIDIG_init_comm(MPIR_Comm * comm)
 
-    MPIR_Assert(subcomm_type <= 3);
-    MPIR_Assert(is_localcomm <= 1);
+    MPIR_Assert(MPIR_CONTEXT_READ_FIELD(SUBCOMM, comm->recvcontext_id) <= 3);
+    MPIR_Assert(MPIR_CONTEXT_READ_FIELD(IS_LOCALCOMM, comm->recvcontext_id) <= 1);
```
these can be removed

raffenet reviewed (commented):
Made comments as I went. Just minor stuff.

wesbland commented:
@raffenet I took care of your comments and added 4 more commits that came up with further internal builds. In particular, the last commit is something you should take a look at because it basically turns off warning checking for submodules.

I think it's unfair to punish MPICH for the submodules warnings. They should have their own warning checkers if we care and in some cases, since we don't control them, it's not reasonable to expect them to be clean.

wesbland commented:
I moved all of the `aclocal_cc.m4` commits to the beginning and squashed them where it made sense. I also differentiated between the Intel class compiler and the Intel oneAPI compiler. The later includes `-Wno-unused-labels` while the former does not. Take another quick look at that one before giving it the final 👍 please.

hzhou reviewed (approved):
LGTM
